src/Loess.jl

```diff
@@ -179,48 +195,147 @@ end
+on the calculations, see the cited reference.
+
+Reference: Cleveland, William S., and Eric Grosse. "Computational methods
+for local regression." Statistics and computing 1, no. 1 (1991): 47-62.
+"""
+function predict(
+    model::LoessModel,
+    x::AbstractVecOrMat = modelmatrix(model);
+    interval::Union{Symbol, Nothing}=nothing,
+    level::Real=0.95
+)
     if size(model.xs, 2) > 1
         throw(ArgumentError("multivariate blending not yet implemented"))
     end
 
-    return [predict(model, z) for z in zs]
-end
+    if interval !== nothing && interval !== :confidence
+        if interval === :prediction
+            throw(ArgumentError("predictions not implemented. If you know how to compute them then please file an issue and explain how."))
+        end
+        throw(ArgumentError("interval must be either :prediction or :confidence but was $interval"))
+    end
 
-function predict(model::LoessModel, zs::AbstractMatrix)
-    if size(model.xs, 2) != size(zs, 2)
-        throw(DimensionMismatch("number of columns in input matrix must match the number of columns in the model matrix"))
+    if !(0 < level < 1)
+        throw(ArgumentError("level must be between zero and one but was $level"))
     end
 
-    if size(zs, 2) == 1
-        return predict(model, vec(zs))
+    predictions = [predict(model, _x) for _x in x]
+
+    if interval === nothing
+        return predictions
     else
-        return [predict(model, row) for row in eachrow(zs)]
+        # see Cleveland and Grosse 1991 p.50.
+        L = hatmatrix(model)
+        L̄ = L - I
+        L̄L̄ = L̄' * L̄
+        δ₁ = tr(L̄L̄)
+        δ₂ = sum(abs2, L̄L̄)
+        ε̂ = L̄ * model.ys
+        s = sqrt(sum(abs2, ε̂) / δ₁)
+        ρ = δ₁^2 / δ₂
+        qt = quantile(TDist(ρ), (1 + level) / 2)
+        sₓ = [s*sqrt(sum(abs2, _hatmatrix_x(model, _x))) for _x in x]
+        lower = [_x - qt * _sₓ for (_x, _sₓ) in zip(predictions, sₓ)]
+        upper = [_x + qt * _sₓ for (_x, _sₓ) in zip(predictions, sₓ)]
+        return LoessPrediction(predictions, lower, upper, sₓ, δ₁, δ₂, s, ρ)
     end
 end
 
 fitted(model::LoessModel) = predict(model, modelmatrix(model))
 
 residuals(model::LoessModel) = response(model) .- fitted(model)
 
+function hatmatrix(model::LoessModel{T}) where T
+    n = length(model.ys)
+    L = zeros(T, n, n)
+    for (i, r) in enumerate(eachrow(model.xs))
+        z = only(r) # we still only support one predictor
+        L[i, :] = _hatmatrix_x(model, z)
+    end
+    return L
+end
+
+# Compute elements of the hat matrix of `model` at `x`.
+# See Cleveland and Grosse 1991 p. 54.
+function _hatmatrix_x(model::LoessModel{T}, x::Number) where T
+    n = length(model.ys)
+
+    adjacent_verts = traverse(model.kdtree, (T(x),))
+
+    @assert(length(adjacent_verts) == 2)
+    v₁, v₂ = adjacent_verts[1], adjacent_verts[2]
+
+    if x == v₁ || x == v₂
+        Lx = model.hatmatrix_generator[[x]][1, :]
```

Comment:
If it's an inplace function and the type of the cache is changed:
```suggestion
        copyto!(Lx, view(model.hatmatrix_generator[x], 1, :))
```